app/src/main/java/com/alphawallet/app/service/AssetDefinitionService.java

```diff
@@ -346,12 +336,26 @@ public TokenDefinition getAssetDefinition(int chainId, String address)
         return assetDef; // if nothing found use default
     }
 
-    public Single<TokenDefinition> getAssetdefinitionASync(int chainId, final String address)
+    public Single<TokenDefinition> getAssetDefinitionASync(int chainId, final String address)
     {
         if (address == null) return Single.fromCallable(TokenDefinition::new);
         String contractName = address;
         if (contractName.equalsIgnoreCase(tokensService.getCurrentAddress())) contractName = "ethereum";
 
+        // hold until asset definitions have finished loading
+        try
+        {
+            assetLoadingLock.acquire();
```

Comment:
If an InterruptedException occurs, the semaphore is not acquired, so the `release` in finally is wrong.
`release` should be right after acquire() in this usage.